### ISSUE TITLE
graphql playground is accessible outside the internal network

### ISSUE THREAD

**Issue by testxss89** [apollographql/apollo-server#3679]
hello i know that this is not a bug but i searched a lot for response but nothing, i'm pentester and while checking my target i found that the graphql playground is accessible from internet, should this happen ?

**Comment by leroy0211** [comment]
Shouldn't this be your clients responsibility by disabling the playground in production? Or by limiting access to the playground url (if this is a different url)?

**Comment by testxss89** [comment]
@leroy0211 Yep that's what i'm asking about, is this the logic configuration ? should this be accessible outside thier network ? the url is something like this 'internal-graphql.target.com' 
they r using it as api to fetch data ,api ok , but what about this accessible instance , i think the logic config is that url be used by the api and without any index, but as i mention the playground is accessible !
thank you @leroy0211 ,

**Comment by testxss89** [comment]
So it there any risk on this ?

**Comment by k00k** [comment]
@testxss89 It would be up to your client to lock this down properly. Depending on the version of `apollo-server` they are using, the implementation that they are using, and the manner in which they are instantiating it, the playground may or may not be accessible in a production environment. As a pentester, you should report this to your client so they can look into the correct way to lock this down based on their needs.

> So it there any risk on this ?

This depends on too many things. Everything from the queries/mutations present, to the sensitivity of the data exposed. They could have mutations that expose user data, or they could be a public API that just says "Hello World." Regardless, we don't know your client, so it's up to them to understand what a publicly accessible playground means. Generally, it's considered best practice to turn off the playground on production environments. This is now a default in `apollo-server`. Typically, through a playground, one can look at the queries and mutations available on a given graphql server and then attempt to run them. This COULD mean that someone nefarious could make calls to those and do something with the data response. All of this depends on how credentials are handled, and way too many other possibilities to guess at in this reply.

Bottom line is to let your client know that you can access the playground and how so, so that they can make the right call for them.

**Comment by abernix** [comment]
GraphQL Playground by itself is just a browser frontend for a GraphQL server.  If introspection is disabled, it's no more powerful than the `curl` command.  By default, playground and introspection are disabled in production (when the environment variable `NODE_ENV` is set to `production`.  That behavior can be changed using [the steps in our documentation](https://www.apollographql.com/docs/apollo-server/testing/graphql-playground/), but it's up to the individual user to determine the importance of hiding their schema entirely.  If they only want to allow very specific operations, they should consider implementing operation safe-listing or other controls (e.g. auth) to limit access to the graph.
